### ISSUE TITLE
Add type-hints to pysmt code

### ISSUE THREAD

**Issue by enmag** [pysmt/pysmt#780]
Python has added support for type-hints in version [3.5](https://www.python.org/downloads/release/python-350/); which has now reached end-of-life.
 
In my opinion, type-hints make it easier to read the code and also enable a number of static checks.     
What do you think?

If we want to go fancy, there are a number of different formats / libraries that support the definition of contracts (mostly pre-post conditions and invariants e.g. [PEP 316](https://www.python.org/dev/peps/pep-0316/), [deal](https://github.com/life4/deal)).

**Comment by mikand** [comment]
I fully agree that adding type hints would be great! The only reason we do not have them (yet) is that we started pysmt with python 2.7 and we never had time to introduce them afterwards...

**Comment by enmag** [comment]
Here is my first attempt: #782
